### PR TITLE
Add plumed interfaces folder to library

### LINKED ISSUE #248: Inconsistency in n_out channels of committor and PYTORCH_GNN in PLUMED
Greetings!
I am trying to use committor without descriptors in plumed, but get the following error:
```bash
(PytorchGNN.cpp:520) PLMD::colvar::pytorch_gnn::PytorchGNN::PytorchGNN(const
PLMD::ActionOptions&)
The committor model should output two values!
```

Committor gives one output value which is z or q, depending on Sigmoid. Plumed module awaits both:

```bash
When the input PyTorch model
is a committor model, the outputs will assign the zeta value to the first
output (node-0) and the q value to the second output (node-1).
```

Am I missing something, or should I wrap the Committor class to get the proper input for plumed? Using model.ptc from diala gives the same error.

Best regards,
Philipp

### LINKED ISSUE #249: Add PLUMED interfaces section to the library
We should also add the `PLUMED` interfaces to the lib to have a place where to find updated code if under development, especially for those interfaces that are not available in the default `PLUMED` installations (e.g., Kolmogorov bias and GNN-based models)
See also #248

### PR DESCRIPTION
**Pull request by EnricoTrizio**
## Description
This PR adds a folder to the repository to store updated and stable versions of the PLUMED interfaces. 
This fixes #249 and fixes #248

## Todos
  - [x] Revise interface for descriptors-based for MLCVs (i.e. `PytorchModel.cpp`)
  - [x] Revise the interface for descriptors-based for committor-based Kolmogorov bias (i.e. `PytorchModelBias.cpp`)
  - [x] Revise GNN (i.e. `PytorchModelGNN.cpp` and maybe `PytorchModelGNNBias.cpp`) interface for flexibility and consitency with the 2.0 release code
  - [x] Add docstrings in PLUMED format

## Status
- [x] Ready to go

### CODE REVIEW

File: plumed_interfaces/PytorchModelGNN.cpp [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+    "optional",
+    "NL_STRIDE",
+    "The frequency with which we are updating the atoms in the neighbor list"
+  );
+
+  keys.add(
+    "optional",
+    "BUFFER",
+    "Buffer size used in finding active environment atoms"
+  );
```
Comment (Kai-Zhu-2001) [inline plumed_interfaces/PytorchModelGNN.cpp]: The cutoff is already stored in the model and can be automatically read by the interface. However, the buffer still needs to be manually specified by the user. I think it would be preferable to store the buffer in the model as well, allowing the interface to read it automatically.

File: plumed_interfaces/PytorchModelGNN.cpp [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+  // cell
+  // TODO: now, the box data used by the model file is in unit of Angstrom.
+  // We should warn the users about this default
+  PLMD::Tensor box = getBox();
+  auto cell = torch::zeros({3, 3}, torch_float_dtype);
+  for (int i = 0; i < 3; i++) {
+    for (int j = 0; j < 3; j++)
+      cell[i][j] = box[i][j] * to_ang;
+  }
+  cell = cell.to(device);
```
Comment (Kai-Zhu-2001) [inline plumed_interfaces/PytorchModelGNN.cpp]: I think it would be better to switch to the following implementation:
```
  PLMD::Tensor box = getBox();
  std::vector<float> cell_vector(9);
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 3; j++)
      cell_vector[i * 3 + j] = box[i][j] * to_ang;
  }
  torch::Tensor cell = torch::from_blob(
    cell_vector.data(),
    9,
    torch::TensorOptions().dtype(torch::kFloat32)
  );
  cell = cell.to(device).to(torch_float_dtype);
  cell = cell.reshape({3, 3});
```

File: plumed_interfaces/PytorchModelGNN.cpp (hunk 1 above)
Comment (EnricoTrizio) [reply plumed_interfaces/PytorchModelGNN.cpp]: I also thought about it, the problem is that the buffer size is not an attribute of the model but is defined when the graphs are initialized in the dataset, so in case we would need to modify a few things to avoid confusion. I don't know what is best, @luigibonati what do you think?

File: plumed_interfaces/PytorchModelGNN.cpp [hunk 3]
```diff
@@ -0,0 +1,980 @@
...
+  // We should warn the users about this default
+  auto positions = torch::empty({n_atoms, 3}, torch_float_dtype);
+  #pragma omp parallel for num_threads(n_threads)
+  for (int i = 0; i < n_atoms; i++) {
+    int index = atom_list_active[i];
+    positions[i][0] = x_local[index][0] * to_ang;
+    positions[i][1] = x_local[index][1] * to_ang;
+    positions[i][2] = x_local[index][2] * to_ang;
+  }
+  positions = positions.to(device);
```
[the remark this answers is not in the document]
Comment (EnricoTrizio) [reply plumed_interfaces/PytorchModelGNN.cpp]: okk

File: plumed_interfaces/PytorchKolmogorovBias.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+  torch::Tensor t_grad_output_bias;
+public:
+  explicit PytorchModelBias(const ActionOptions&);
+  void calculate();
+  static void registerKeywords(Keywords& keys);
+
+  std::vector<float> tensor_to_vector(const torch::Tensor& x);
+};
+
+PLUMED_REGISTER_ACTION(PytorchModelBias,"PYTORCH_MODEL_BIAS")
```
Comment (luigibonati) [inline plumed_interfaces/PytorchKolmogorovBias.cpp]: shouldn't be called PYTORCH_KOLMOGOROV_BIAS?

File: plumed_interfaces/PytorchModelGNN.cpp (hunk 1 above)
Comment (luigibonati) [reply plumed_interfaces/PytorchModelGNN.cpp:265]: I think the key question is whether the buffer is a runtime parameter or fixed by construction. If it can be changed at runtime, then it makes sense to keep it as a keyword in PLUMED. If instead it is tied to how the graph was built during training, I think it should be read from the model.

Of course, in the GNN case the model and the dataset are not independent, since the graph is already constructed in the dataset. One possible approach could be to propagate the relevant dataset attributes into the model at compilation time. For example, when calling to_torchscript, we could also pass the dataset/dataloader and store in the model any additional properties that are required for inference but were originally defined in the dataset.

**Comment by luigibonati** [comment]
I think we could improve the documentation on how to use the interfaces.

- I would add a README file in the github folder, pointing to the documentation. 
- Either in the README or in the doc, we should also mention how to use them at runtime via LOAD. furthermore, since some of the registered keyword are different, shall we mention their names in the documentation alongside the *.cpp names?
- Each folder in the test input should also contain a short README with the commands to run them. In this way one can look at the test also to know how to use them.

Plus: it could be worth explaining the expected input of the GNN interface? Is it the same as the `Configuration` object?

File: plumed_interfaces/PytorchModelGNN.cpp (hunk 1 above)
Comment (EnricoTrizio) [reply plumed_interfaces/PytorchModelGNN.cpp:265]: In principle, it can be changed. I don't know how common that would be anyway, so maybe we can directly take it from the model.

For having the buffer in the model, we can directly have that GNN models can be either initialized key by key or by passing the `dictionary.metadata`, where all those parameters are stored.
Then, in `to_torchscript`, we can add a warning and the possibility of passing the `dictionary.metadata` if needed

What do you think?

File: plumed_interfaces/PytorchModelGNN.cpp [hunk 5]
```diff
@@ -0,0 +1,1013 @@
...
+        cell_inv, positions.transpose(1, 0)
+      );
+      auto deltas = (
+        torch::index_select(positions_s, 1, edge_index[0])
+        - torch::index_select(positions_s, 1, edge_index[1])
+      );
+      unit_shifts = torch::round(deltas);
+      shifts = torch::matmul(
+        cell.transpose(1, 0), unit_shifts
+      ).transpose(1, 0);
```
Comment (Kai-Zhu-2001) [inline plumed_interfaces/PytorchModelGNN.cpp:823]: The dimension of `unit_shifts` is incorrect. It should be transposed as follows:
```
unit_shifts = torch::round(deltas).transpose(1, 0);
shifts = torch::matmul(unit_shifts, cell);
```
